File: src/main/java/com/ocadotechnology/gembus/test/rearranger/Rearranger.java
/*
 * Copyright © 2020 Ocado (marian.jureczko@ocado.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ocadotechnology.gembus.test.rearranger;

import java.util.Map;
import java.util.Set;
import java.util.function.Supplier;

/**
 * Test utility that produces a shallow copy of an existing instance while allowing selected
 * field values to be overridden. This Java implementation plays a role analogous to the Kotlin
 * {@code com.ocadotechnology.gembus.test.rearangerkt.Rearranger} singleton, but uses a simple
 * {@code Map<String, Supplier<?>>} based override syntax instead of Kotlin property references.
 * <p>
 * Supports both Java Records and regular (POJO) classes from a single entry point.</li>
 * <p>
 * <p><strong>Usage example (POJO):</strong>
 * <pre>{@code
 * PojoClass original = some(PojoClass.class);
 * PojoClass modified = Rearranger.copy(original, Map.of(
 *     "name", () -> "Overridden Name",
 *     "number", () -> 42
 * ));
 * }</pre>
 * <strong>Usage example (Record):</strong>
 * <pre>{@code
 * MyRecord original = new MyRecord("Alice", 10);
 * MyRecord modified = Rearranger.copy(original, Map.of(
 *     "name", () -> "Bob"
 * ));
 * }</pre>
 * <p>
 * Notes / Limitations:
 * <ul>
 *   <li>This is a shallow copy: referenced mutable objects are not cloned.</li>
 *   <li>Each supplier is invoked exactly once (during copy construction).</li>
 *   <li>Overrides must reference existing field / record component names; otherwise an {@link IllegalArgumentException} is thrown.</li>
 *   <li>Reflection may set accessibility to true to reach non-public members (for non-record classes).</li>
 * </ul>
 * <p>
 *
 * @see com.ocadotechnology.gembus.test.rearrangerkt.Rearranger Kotlin counterpart using property references
 */
public class Rearranger {

    /**
     * Creates a shallow copy of the provided {@code object}, applying zero or more field overrides.
     *
     * @param object    the original instance to copy; must not be {@code null}
     * @param overrides map whose keys are field (or record component) names and whose values are suppliers
     *                  producing the replacement values. A missing key means the original value is retained.
     * @param <T>       the static type of the copied object
     * @return a new instance of {@code T} with supplied overrides applied (shallow copy)
     * @throws IllegalArgumentException if an override refers to a field/component that does not exist on the class
     * @throws RuntimeException         wrapping reflective failure encountered during copying
     */
    public static <T> T copy(T object, Map<String, Supplier<?>> overrides) {
        try {
            if (object.getClass().isRecord()) {
                return RecordRearranger.copyRecord(object, overrides);
            } else {
                return ObjectRearranger.copyObject(object, overrides);
            }
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }

    static <T> void validateOverrides(Map<String, Supplier<?>> overrides, Set<String> fieldNames, String className) {
        overrides.keySet().forEach(fieldToOverride -> {
            if (!fieldNames.contains(fieldToOverride)) {
                throw new IllegalArgumentException("Failed to override field " + fieldToOverride + " in class " + className + ". Field not found.");
            }
        });
    }
}

File: src/main/kotlin/com/ocadotechnology/gembus/test/Rearranger.kt
/*
 * Copyright © 2020 Ocado (marian.jureczko@ocado.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.ocadotechnology.gembus.test

import org.objenesis.Objenesis
import org.objenesis.ObjenesisStd
import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.KMutableProperty1
import kotlin.reflect.KProperty1
import kotlin.reflect.full.memberProperties
import kotlin.reflect.jvm.isAccessible
import kotlin.reflect.jvm.javaField

/**
 * Utility for creating modified copies of test data objects regardless of their mutability.
 * Allows creation of new instances with selected properties overridden without having to
 * manually implement copy constructors or builder patterns.
 */
object Rearranger {

    /**
     * Creates a copy of the given instance, allowing to override some of its properties.
     * Example usage:
     * ```
     * val original = SomeClass(id = 123, name = "Original")
     * val modified = Rearranger.copy(original) {
     *  SomeClass::name set "Modified"
     * }
     * ```
     */
    inline fun <reified T : Any> copy(
        instance: T,
        block: CopyScope<T>.() -> Unit
    ): T {
        val scope = CopyScope(instance)
        scope.block()
        return scope.build()
    }

    class CopyScope<T : Any>(private val original: T) {
        private val overrides = mutableMapOf<KProperty1<T, *>, Any?>()
        private val objenesis: Objenesis = ObjenesisStd()

        /**
         * Override a property value in the copy: `SomeClass::foo set newValue`
         */
        infix fun <R> KProperty1<T, R>.set(value: R) {
            overrides[this] = value
        }

        fun build(): T {
            val props = original::class.memberProperties
                .filterIsInstance<KProperty1<T, Any?>>()

            val matchingCtor = findMatchingConstructor(original::class, props)

            if (matchingCtor != null) {
                return createCopyWithOverridesUsingContructor(matchingCtor, props)
            } else {
                val toFillWithValues = objenesis.newInstance(original::class.java)
                setValuesInCopiedInstance(props, toFillWithValues)
                return toFillWithValues
            }
        }

        private fun findMatchingConstructor(
            kClass: KClass<out T>,
            props: List<KProperty1<T, Any?>>
        ) = kClass.constructors.firstOrNull { ctor ->
            val paramNames = ctor.parameters.mapNotNull { it.name }.toSet()
            val propNames = props.map { it.name }.toSet()
            propNames.all { it in paramNames }
        }

        private fun createCopyWithOverridesUsingContructor(
            matchingCtor: KFunction<T>,
            props: List<KProperty1<T, Any?>>
        ): T {
            val args = matchingCtor.parameters.associateWith { param ->
                val prop = props.first { it.name == param.name }
                overrides[prop] ?: prop.get(original)
            }
            @Suppress("UNCHECKED_CAST")
            return matchingCtor.callBy(args) as T
        }

        private fun setValuesInCopiedInstance(props: List<KProperty1<T, Any?>>, toFillWithValues: T) {
            props.forEach { prop ->
                val value = overrides[prop] ?: prop.get(original)
                setFieldValue(prop, toFillWithValues, value)
            }
        }

        private fun setFieldValue(fieldType: KProperty1<T, Any?>, field: T, value: Any?) {
            when (fieldType) {
                is KMutableProperty1<T, Any?> -> {
                    fieldType.isAccessible = true
                    fieldType.set(field, value)
                }

                else -> {
                    fieldType.javaField?.apply {
                        isAccessible = true
                        set(field, value)
                    }
                }
            }
        }
    }
}